### PR TITLE
feat: adding scope to budget limits table

### PR DESCRIPTION
**Pull request by roroghost17**
## Summary

Model configs previously applied globally to all traffic. This PR introduces a `scope` / `scope_id` system that allows model-level rate limits and budgets to be pinned to a specific virtual key, so per-VK model limits can be enforced independently of (and in addition to) the global model limits.

## Changes

- Added `scope` (default `"global"`) and `scope_id` (nullable) columns to `governance_model_configs` via a new idempotent migration. Existing rows are backfilled to `"global"` and the unique index is swapped from `(model_name, provider)` to `(scope, scope_id, model_name, provider)`.
- Introduced `ModelConfigScopeGlobal` and `ModelConfigScopeVirtualKey` constants and a `BeforeSave` hook that validates and normalises scope/scope_id on every write.
- Extended `GetModelConfig` to accept `scope` and `scopeID` as lookup parameters so the identity check on create is scope-aware.
- Added `CheckVirtualKeyScopedModelBudget` / `CheckVirtualKeyScopedModelRateLimit` and their corresponding `UpdateVirtualKeyScopedModel*UsageInMemory` methods to `GovernanceStore`. These are wired into `EvaluateVirtualKeyRequest` (resolver) and `UsageTracker.UpdateUsage` so scoped limits are both enforced pre-request and incremented post-response.
- In-memory store keys are namespaced by scope via `modelConfigStoreKey`, preventing global and scoped configs from colliding. `DeleteVirtualKeyInMemory` now evicts scoped model configs (and their owned budgets/rate-limits) when a VK is removed; `DeleteVirtualKey` does the same on the DB side.
- `GenerateModelConfigHash` now includes `scope` and `scope_id` so config.json ↔ DB drift detection works correctly for scoped configs.
- The `getModelConfigs` HTTP handler gains a `?from_memory=true` shortcut and enriches all responses with a transient `scope_name` field (resolved VK name) for UI display.
- `CreateModelConfigRequest` accepts `scope` / `scope_id`; the handler validates the scope value, enforces that `virtual_key` scope references an existing VK, and returns scope-aware conflict messages.
- UI: the model-limit sheet gains a Scope selector and a virtual-key combobox (shown only for the `virtual_key` scope). The model-limits table gains a Scope column. `getScopeLabel` is extracted to a shared `ui/lib/utils/labels.ts` and the routing-rules views are updated to import from there.
- `.zed/` added to `.gitignore`.

## Type of change

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [x] Transports (HTTP)
- [ ] Providers/Integrations
- [x] Plugins
- [x] UI (React)
- [ ] Docs

## How to test

```sh
# Core/Transports
go test ./framework/configstore/... ./plugins/governance/...

# UI
cd ui
pnpm i
pnpm build
```

1. Start the server against a fresh or existing DB — the migration runs automatically and backfills existing model configs to `scope = "global"`.
2. Create a virtual key, then create a model config with `scope = "virtual_key"` and `scope_id = <vk_id>` via the UI or `POST /api/governance/model-configs`.
3. Send requests using that virtual key for the configured model and verify the scoped rate limit / budget is enforced independently of the global model config.
4. Delete the virtual key and confirm the scoped model config and its owned budget/rate-limit rows are removed.

## Breaking changes

- [ ] Yes
- [x] No

The migration is additive and fully backward-compatible. Existing model configs default to `"global"` scope and behave identically to before.

## Security considerations

`scope_id` for the `virtual_key` scope is validated against the VK table on creation, preventing configs from being attached to non-existent keys. `scope_id` carries no FK constraint in the DB, so the explicit cascade cleanup on VK deletion is required to avoid orphaned rows — this is implemented both in the DB layer and the in-memory store.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [ ] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [ ] I verified the CI pipeline passes locally if applicable

## Summary by CodeRabbit

* **New Features**
  * Model limits and rate limits can be scoped to individual virtual keys; UI supports selecting scope and picking a virtual key. API accepts scope/scope_id on create.

* **UI**
  * Model limits table shows a Scope column and friendly scope labels; forms include Scope + Virtual Key selector.

* **Behavior / Bug Fixes**
  * Deleting a virtual key removes its scoped model configs and related owned records. Conflicts now consider scope/scope_id.

* **Tests & Migrations**
  * Added tests and a migration to introduce and backfill model-config scoping.

### CODE REVIEW

**Comment by roroghost17** [comment]
* **#3940** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3940?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#3960](https://github.com/maximhq/bifrost/pull/3960) <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3960?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3962](https://github.com/maximhq/bifrost/pull/3962) <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3962?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3939** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3939?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3938** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3938?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3937** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3937?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3937?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `dev`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: framework/configstore/migrations.go [hunk 1]
```diff
@@ -3832,6 +3835,83 @@ func migrationAddProviderGovernanceColumns(ctx context.Context, db *gorm.DB) err
...
+					return fmt.Errorf("failed to create idx_model_scope_provider: %w", err)
+				}
+			}
+			// Drop the now-superseded (model_name, provider) unique index.
+			if migrator.HasIndex(modelConfig, "idx_model_provider") {
+				if err := migrator.DropIndex(modelConfig, "idx_model_provider"); err != nil {
+					return fmt.Errorf("failed to drop idx_model_provider: %w", err)
+				}
+			}
+			return nil
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations.go:3884]: The governance_model_configs table is empty at the point this migration runs — all provider-level and VK-level model configs are populated by later migrations (provider governance and VK migrations) that execute after this one. The ACCESS EXCLUSIVE lock window is therefore milliseconds on an empty or near-empty table, so there is no meaningful contention risk in practice. We'll keep the GORM migrator path for dialect safety.

File: framework/configstore/migrations.go [hunk 2]
```diff
@@ -3832,6 +3835,83 @@ func migrationAddProviderGovernanceColumns(ctx context.Context, db *gorm.DB) err
...
+			// unenforced. CreateIndex reads the struct tags so it is dialect-safe.
+			if !migrator.HasIndex(modelConfig, "idx_model_scope_provider") {
+				if err := migrator.CreateIndex(modelConfig, "idx_model_scope_provider"); err != nil {
+					return fmt.Errorf("failed to create idx_model_scope_provider: %w", err)
+				}
+			}
+			// Drop the now-superseded (model_name, provider) unique index.
+			if migrator.HasIndex(modelConfig, "idx_model_provider") {
+				if err := migrator.DropIndex(modelConfig, "idx_model_provider"); err != nil {
+					return fmt.Errorf("failed to drop idx_model_provider: %w", err)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations.go:3881]: The governance_model_configs table is empty at the point this migration runs — all provider-level and VK-level model configs are populated by later migrations (provider governance and VK migrations) that execute after this one. The ACCESS EXCLUSIVE lock window is therefore milliseconds on an empty or near-empty table, so there is no meaningful contention risk in practice. We'll keep the GORM migrator path for dialect safety.

File: framework/configstore/migrations_test.go [hunk 3]
```diff
@@ -2381,3 +2381,59 @@ func assertNoCorruptedFKReferences(t *testing.T, db *gorm.DB) {
...
+			updated_at datetime NOT NULL
+		)
+	`).Error)
+	require.NoError(t, db.Exec(`CREATE UNIQUE INDEX idx_model_provider ON governance_model_configs (model_name, provider)`).Error)
+
+	now := time.Now()
+	require.NoError(t, db.Exec(`
+		INSERT INTO governance_model_configs (id, model_name, created_at, updated_at)
+		VALUES (?, ?, ?, ?)
+	`, "mc-existing", "gpt-4", now, now).Error)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations_test.go:2412]: Acknowledged — this is a nitpick and the existing test covers the primary invariant (backfill to global scope, index swap). Additional coverage for batch backfill and constraint boundary cases can be added in a follow-up once the full scope feature lands.

File: framework/configstore/migrations.go [hunk 4]
```diff
@@ -3832,6 +3835,83 @@ func migrationAddProviderGovernanceColumns(ctx context.Context, db *gorm.DB) err
...
+				}
+			}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations.go:3883]: Valid observation. In practice, duplicate global rows are also prevented at the application layer: the handler calls GetModelConfig before any insert and rejects duplicates with a 409. This table is only written via the admin UI (low concurrency, no bulk-insert paths), so a concurrent-insert race through that check is extremely unlikely. Hardening with a non-NULL sentinel for scope_id on global rows is a worthwhile follow-up but out of scope here.

File: plugins/governance/store.go [hunk 5]
```diff
@@ -1003,27 +1009,64 @@ func (gs *LocalGovernanceStore) CheckProviderRateLimit(ctx context.Context, requ
...
+	}
+	return []modelConfigScope{{name: configstoreTables.ModelConfigScopeVirtualKey, id: vk.ID}}
+}
+
+// findModelOnlyConfig looks up a global, model-only config (no provider) with cross-provider
+// model name normalization. Returns the matching config and the display name for error messages.
 func (gs *LocalGovernanceStore) findModelOnlyConfig(ctx context.Context, model string) (*configstoreTables.TableModelConfig, string) {
-	// If modelMatcher is available, try normalized base model name first (cross-provider matching)
+	return gs.findScopedModelOnlyConfig(ctx, configstoreTables.ModelConfigScopeGlobal, "", model)
+}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go]: This was a stale analysis artifact from the stack build ordering at review time. configstoreTables is imported as github.com/maximhq/bifrost/framework/configstore/tables and ModelConfigScopeGlobal is defined there — the package builds cleanly now.

File: plugins/governance/store.go [hunk 6]
```diff
@@ -1224,6 +1267,96 @@ func (gs *LocalGovernanceStore) CheckModelRateLimit(ctx context.Context, request
...
+		}
+		// Always check scoped model-only config (cross-provider normalization preserved)
+		if mc, configKey := gs.findScopedModelOnlyConfig(ctx, scope.name, scope.id, model); mc != nil && mc.RateLimitID != nil {
+			if rateLimit := gs.LoadRateLimit(ctx, *mc.RateLimitID); rateLimit != nil {
+				ewKey := fmt.Sprintf("Model:%s:%s:%s", configKey, scope.name, scope.id)
+				entityWiseRateLimits[ewKey] = []*configstoreTables.TableRateLimit{rateLimit}
+			}
+		}
+	}
+	return gs.CheckRateLimit(ctx, entityWiseRateLimits, tokensBaselines, requestsBaselines)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go:1351]: Already handled — CollectApplicableGovernanceIDs now traverses nonGlobalModelConfigScopeChain(vk) and calls addModelConfigIDs for each scoped model config (lines 2512–2516), so VK-scoped budget and rate-limit IDs are included in the collector.

[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go:1351]: The CheckVirtualKeyScopedModelRateLimit function referenced here was refactored in the upper stack into the generic CheckScopedModelRateLimit(scope, scopeID, ...), which uses collectModelConfigsFor to return an ordered slice — each entry gets a unique key in entityWiseRateLimits so the two-entry scenario described is resolved. The early-exit behavior in CheckRateLimit itself is pre-existing across all callers and can be addressed separately.

File: ui/app/workspace/model-limits/views/modelLimitSheet.tsx [hunk 7]
```diff
@@ -60,6 +69,7 @@ export default function ModelLimitSheet({ modelConfig, onSave, onCancel }: Model
 	};
 
 	const { data: providersData } = useGetProvidersQuery();
+	const { data: vksData = { virtual_keys: [] } } = useGetVirtualKeysQuery();
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply ui/app/workspace/model-limits/views/modelLimitSheet.tsx:72]: Agreed with your own assessment — always-fetching avoids a visible loading delay when the user switches to virtual_key scope, and VK lists are typically small. Leaving as-is.

File: framework/configstore/migrations.go [hunk 8]
```diff
@@ -3832,6 +3835,83 @@ func migrationAddProviderGovernanceColumns(ctx context.Context, db *gorm.DB) err
...
+					return fmt.Errorf("failed to drop scope_id column: %w", err)
+				}
+			}
+			if migrator.HasColumn(modelConfig, "scope") {
+				if err := migrator.DropColumn(modelConfig, "scope"); err != nil {
+					return fmt.Errorf("failed to drop scope column: %w", err)
+				}
+			}
+			return nil
+		},
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations.go:3888]: Fixed in PR #3940 (top of stack) — rollback replaced with a non-rollbackable error.

File: plugins/governance/tracker.go [hunk 9]
```diff
@@ -124,6 +124,19 @@ func (t *UsageTracker) UpdateUsage(ctx context.Context, update *UsageUpdate) {
...
+	if update.Model != "" {
+		if err := t.store.UpdateVirtualKeyScopedModelRateLimitUsageInMemory(ctx, vk, update.Model, update.Provider, update.TokensUsed, shouldUpdateTokens, shouldUpdateRequests); err != nil {
+			t.logger.Error("failed to update scoped model rate limit usage for VK %s: %v", vk.ID, err)
+		}
+		if shouldUpdateBudget && update.Cost > 0 {
+			if err := t.store.UpdateVirtualKeyScopedModelBudgetUsageInMemory(ctx, vk, update.Model, update.Provider, update.Cost); err != nil {
+				t.logger.Error("failed to update scoped model budget usage for VK %s: %v", vk.ID, err)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/tracker.go:138]: Fixed in PR #3940 (top of stack) — removed the Provider != "" guard from the global model rate-limit/budget update blocks in tracker.go so provider-less model requests also update global model-only counters.

File: framework/configstore/rdb.go [hunk 10]
```diff
@@ -2915,6 +2915,32 @@ func (s *RDBConfigStore) DeleteVirtualKey(ctx context.Context, id string, tx ...
...
+			}
+			if mc.RateLimitID != nil {
+				rateLimitIDs = append(rateLimitIDs, *mc.RateLimitID)
+			}
+		}
+		if err := txDB.WithContext(ctx).
+			Where("scope = ? AND scope_id = ?", tables.ModelConfigScopeVirtualKey, id).
+			Delete(&tables.TableModelConfig{}).Error; err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/rdb.go:2984]: Fixed in this PR — the refactored code at lines 2983–2992 collects budgetIDs/rateLimitIDs, deletes the model configs first, then batch-deletes the owned budgets (id IN ?) and rate-limits (id IN ?). No rows are leaked.

[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/rdb.go:2984]: Fixed in this PR — the refactored code collects IDs, deletes model configs first, then batch-deletes owned budgets and rate-limits. See the IN ? batch deletes after the TableModelConfig delete.

File: framework/configstore/rdb_test.go [hunk 11]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"encoding/json"
 	"testing"
+	"time"
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/rdb_test.go:8]: The time import is still used at line 698 (now := time.Now() for the provider/budget/rate-limit setup). The comment was based on the diff removing some time.Now() usages, but other usages remain — the package builds without error.

File: framework/configstore/tables/modelconfig.go [hunk 12]
```diff
@@ -8,14 +8,40 @@ import (
...
-	ModelName   string  `gorm:"type:varchar(255);not null;uniqueIndex:idx_model_provider" json:"model_name"`
-	Provider    *string `gorm:"type:varchar(50);uniqueIndex:idx_model_provider" json:"provider,omitempty"` // Optional provider, nullable
+	ID        string  `gorm:"primaryKey;type:varchar(255)" json:"id"`
+	ModelName string  `gorm:"type:varchar(255);not null;uniqueIndex:idx_model_scope_provider,priority:3" json:"model_name"`
+	Provider  *string `gorm:"type:varchar(50);uniqueIndex:idx_model_scope_provider,priority:4" json:"provider,omitempty"` // Optional provider, nullable
+	// Scope determines where this config applies: "global" (default) or "virtual_key".
+	Scope string `gorm:"type:varchar(50);not null;default:'global';uniqueIndex:idx_model_scope_provider,priority:1" json:"scope"`
+	// ScopeID is the target of a non-global scope (e.g. the virtual key ID). NULL for global.
+	ScopeID     *string `gorm:"type:varchar(255);uniqueIndex:idx_model_scope_provider,priority:2" json:"scope_id,omitempty"`
 	BudgetID    *string `gorm:"type:varchar(255);index:idx_model_config_budget" json:"budget_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/tables/modelconfig.go:37]: Addressed in the 3333354219 thread above — app-layer uniqueness check via GetModelConfig prevents duplicates at the API level, and a sentinel-based solution is tracked as a follow-up.

File: plugins/governance/modelprovidergovernance_test.go [hunk 13]
```diff
@@ -2100,3 +2100,186 @@ func TestStore_CheckModelBudget_NoCatalog_NoMatch(t *testing.T) {
...
+	require.NoError(t, err)
+
+	decision, err := store.CheckVirtualKeyScopedModelBudget(context.Background(), vk, &EvaluationRequest{Model: "gpt-4", Provider: schemas.OpenAI}, nil)
+	assert.NoError(t, err, "Scoped check must not pick up the global config")
+	assert.Equal(t, DecisionAllow, decision)
+
+	// Sanity: the global model check DOES still catch the exceeded global budget.
+	_, gErr := store.CheckModelBudget(context.Background(), &EvaluationRequest{Model: "gpt-4", Provider: schemas.OpenAI}, nil)
+	assert.Error(t, gErr, "Global model check should catch the exceeded global budget")
+}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/modelprovidergovernance_test.go:2199]: Valid suggestion. Provider-specific scoped model config coverage can be added as a follow-up once the full scope feature lands.

File: plugins/governance/modelprovidergovernance_test.go [hunk 14]
```diff
@@ -2100,3 +2100,186 @@ func TestStore_CheckModelBudget_NoCatalog_NoMatch(t *testing.T) {
...
+	store, err := NewLocalGovernanceStore(context.Background(), logger, nil, &configstore.GovernanceConfig{
+		ModelConfigs: []configstoreTables.TableModelConfig{*mc},
+		RateLimits:   []configstoreTables.TableRateLimit{*rateLimit},
+	}, nil)
+	require.NoError(t, err)
+
+	decision, err := store.CheckVirtualKeyScopedModelRateLimit(context.Background(), vk, &EvaluationRequest{Model: "gpt-4", Provider: schemas.OpenAI}, nil, nil)
+	assert.NoError(t, err)
+	assert.Equal(t, DecisionAllow, decision)
+}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/modelprovidergovernance_test.go:2231]: Valid suggestion. A request-limit exceeded test can be added as a follow-up alongside the broader scoped rate-limit test expansion.

File: plugins/governance/test_utils.go [hunk 15]
```diff
@@ -263,6 +263,14 @@ func buildModelConfig(id, modelName string, provider *string, budget *configstor
 	return mc
 }
 
+// buildVKScopedModelConfig builds a model config scoped to a specific virtual key.
+func buildVKScopedModelConfig(id, modelName string, provider *string, vkID string, budget *configstoreTables.TableBudget, rateLimit *configstoreTables.TableRateLimit) *configstoreTables.TableModelConfig {
+	mc := buildModelConfig(id, modelName, provider, budget, rateLimit)
+	mc.Scope = configstoreTables.ModelConfigScopeVirtualKey
+	mc.ScopeID = &vkID
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/test_utils.go:270]: Fixed in PR #3940 (top of stack) — updated to use schemas.Ptr consistently.

File: plugins/governance/store.go [hunk 16]
```diff
@@ -2913,16 +3134,22 @@ func (gs *LocalGovernanceStore) UpdateModelConfigInMemory(ctx context.Context, m
...
 		gs.modelConfigs.Store(key, &clone)
 	} else {
-		key := clone.ModelName
+		modelKey := clone.ModelName
 		if gs.modelCatalog != nil {
-			key = gs.modelCatalog.GetBaseModelName(clone.ModelName)
+			modelKey = gs.modelCatalog.GetBaseModelName(clone.ModelName)
 		}
+		key := modelConfigStoreKey(clone.Scope, scopeID, modelKey, nil)
 		gs.modelConfigs.Store(key, &clone)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go:3153]: The function comment says 'Scope/scope_id are part of a config's identity and do not change on update, so this matches the stored key.' The update handler enforces this — identity fields (scope, scope_id, model_name, provider) form the DB unique key and are not mutable on update. There is no case where the derived key diverges from the stored key, so the sweep is unnecessary.

File: transports/bifrost-http/handlers/governance.go [hunk 17]
```diff
@@ -392,11 +392,15 @@ type UpdateCustomerRequest struct {
 type CreateModelConfigRequest struct {
 	ModelName string                  `json:"model_name" validate:"required"`
 	Provider  *string                 `json:"provider,omitempty"` // Optional provider, nil means all providers
+	Scope     string                  `json:"scope,omitempty"`    // Defaults to "global" if not provided
+	ScopeID   *string                 `json:"scope_id,omitempty"` // Required for non-global scopes (e.g. the virtual key ID)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply transports/bifrost-http/handlers/governance.go:396]: Fixed — added scope (string, default "global") and scope_id (string, optional) to the governance.model_configs[] item schema in transports/config.schema.json.

File: transports/bifrost-http/handlers/governance.go [hunk 18]
```diff
@@ -2489,6 +2493,33 @@ func validateBudget(budget *configstoreTables.TableBudget) error {
...
+		}
+		h.enrichModelConfigScopeNames(ctx, modelConfigs)
+		SendJSON(ctx, map[string]any{
+			"model_configs": modelConfigs,
+			"count":         len(modelConfigs),
+			"total_count":   len(modelConfigs),
+			"limit":         len(modelConfigs),
+			"offset":        0,
+		})
+		return
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply transports/bifrost-http/handlers/governance.go:2520]: Fixed — the from_memory path now reads search/limit/offset from query args, applies in-memory filtering and slicing, and returns the correct envelope (count, total_count, limit, offset), matching the paginated DB path.

File: transports/bifrost-http/handlers/governance.go [hunk 19]
```diff
@@ -2489,6 +2493,33 @@ func validateBudget(budget *configstoreTables.TableBudget) error {
...
+		// Copy into a value slice before enriching: never mutate ScopeName on the
+		// pointers returned here, so we don't risk racing the in-memory store regardless
+		// of whether the manager hands back shared pointers or clones.
+		modelConfigs := make([]configstoreTables.TableModelConfig, 0, len(data.ModelConfigs))
+		for _, mc := range data.ModelConfigs {
+			if mc != nil {
+				modelConfigs = append(modelConfigs, *mc)
+			}
+		}
+		h.enrichModelConfigScopeNames(ctx, modelConfigs)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply transports/bifrost-http/handlers/governance.go:2512]: Fixed — the from_memory path now deep-copies each model config: top-level struct copy + separate copies of Budgets slice, Budget pointer, and RateLimit pointer before passing to enrichModelConfigScopeNames and serializing. No nested pointers alias live governance state.

**Comment by akshaydeo** [comment]
### Merge activity

* **Jun 2, 1:06 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/3937).
* **Jun 2, 1:07 PM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/3937).
